stan/math/rev/core/var.hpp

```diff
@@ -63,185 +74,77 @@ class var {
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  var(long x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  var(unsigned char x)  // NOLINT(runtime/explicit)
-      : vi_(new vari(static_cast<double>(x), false)) {}
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  // NOLINTNEXTLINE
-  var(unsigned short x) : vi_(new vari(static_cast<double>(x), false)) {}
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  // NOLINTNEXTLINE
-  var(unsigned int x)
-      : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
+   * Construct a variable from a pointer to a variable implementation.
+   * @tparam S The type in the vari_value pointer that has the same `value_type`
+   *  as this `var_value`. For integral types a `vari_value<Integral>` is the
+   * same as `vari_value<double>` so can be `reinterpret_cast` without a copy.
+   * @param vi A vari_value pointer.
    */
-  // NOLINTNEXTLINE
-  var(unsigned long x)
-      : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-#ifdef _WIN64
-  // these two ctors are for Win64 to enable 64-bit signed
-  // and unsigned integers, because long and unsigned long
-  // are still 32-bit
+  template <typename S, require_vari_convertible_t<S>* = nullptr>
+  var_value(vari_value<S>* vi)  // NOLINT
+      : vi_(vi) {}
 
   /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
+   * Construct a variable from a pointer to a variable implementation when.
+   *  the `value_type`s differ. This forces a copy of the vari
+   * @tparam S the type in the vari_value* that does not have the same type
+   *  as the `value_type` of this `var_value`.
+   * @param vi A `vari_value`.
    */
-  var(size_t x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
+  template <typename S, require_not_vari_convertible_t<S>* = nullptr>
+  var_value(vari_value<S>* vi) : vi_(new vari_type(*vi)) {}  // NOLINT
```

Comment:
What does this accept that `var_value(S x)` doesn't accept?

Reply:
It's not in what it accepts, it's the little `*` in `vi_(new vari_type(*vi))`. That makes a new vari based on the values of `vi`. So it's used for like 
```
var_value<double>(new vari_value<long double>(1.0))
```

Where we need to make a copy of that vari (since `vari_value<double>` and `vari_value<long double>` are not pointer convertible).

`var_value(S x)` also allocates the new vari on the nochain stack with `vi_(new vari_type(x, false))`

Reply:
Presumably we don't want to unpromote a type though.

Reply:
What do you mean by unpromote? Like we don't want to cast a `vari_value` of one type to another that changes the floating point storage (that's the only time this is called)? tbh I'm not sure if we want to have that happen, but at this time I don't see a good reason to stop it from happening. Could something bad happen here?

Reply:
Well we end up with a `var_value<double>` when we started with a `vari_value<long double>`. That looks like a demotion from `long double` to `double`.

Reply:
Ah, yes I see what you mean now. Yeah idk. I could see instances of wanting to promote / demote the floating point type and if at some point in time we go, "This was a bad idea" we can just remove it. How do you feel about that?

Reply:
I don't know where the regular C-style 1-step casts won't handle this. Like anything that can get cast to a double will, and we're good.

I don't think we should allow something like `var_value<double>(vari_value<float>(1.0f))` until we're ready to define a mixed precision autodiff.

Reply:
> until we're ready to define a mixed precision autodiff.

I don't wanna do this waaaah